api/replica.go

```diff
@@ -14,20 +19,98 @@ type Replica struct {
 	handler   http.Handler
 	tlsConfig *tls.Config
 	primary   string
+	// the address of this replica
+	addr string
+	// mu is the mutex used for locking the primary.
+	mu sync.RWMutex
+	// primaryWait is used to block execution of ServeHTTP until the primary is
+	// ready.
+	primaryWait *sync.Cond
+
+	// capture the hijack function for easier unit testing
+	hijack func(*tls.Config, string, http.ResponseWriter, *http.Request) error
 }
 
 // NewReplica creates a new API replica.
-func NewReplica(handler http.Handler, tlsConfig *tls.Config) *Replica {
-	return &Replica{
+func NewReplica(handler http.Handler, tlsConfig *tls.Config, addr string) *Replica {
+	r := &Replica{
 		handler:   handler,
 		tlsConfig: tlsConfig,
+		addr:      addr,
+		hijack:    hijack,
 	}
+	// This seems to be a big confusing, so here's the explanation:
```

Comment:
Not sure if this was because of my comment, but most of the RW lock and condition variable is pretty standard stuff around synchronization. While I'm not sure if its needed, I'm OK leaving it in there.

My question was more around how golang's mutexes and condition variables were implemented and I didn't realize that the condition variable was only associated with the reader lock from the RW lock.

Reply:
Ah but my comment did answer your question, didn't it? The goal of the comment was to answer every question anyone could ever have.

This was supposed to say  "a bit confusion" and not "a big confusing" but the typo is small enough and the result coherent enough that i don't really feel like pushing up another change, if you don't want to make me.

Reply:
Yep, its fine.